test/test.js

```diff
@@ -63,9 +277,149 @@ describe("ERC721RExample", function () {
+  it("[Owner:Revert] Owner can not withdraw when `Refund period not over`", async function () {
+    await expect(erc721RExample.connect(owner).withdraw()).to.revertedWith(
+      "Refund period not over"
+    );
+  });
+
+  it("[Owner] Owner can withdraw after refundEndTime", async function () {
+    const refundEndTime = await erc721RExample.refundEndTime();
+
+    await erc721RExample
+      .connect(account2)
+      .publicSaleMint(1, { value: parseEther(MINT_PRICE) });
+
+    // refund period is over, just refundEndTime + 1 second.
+    await erc721RExample.provider.send("evm_setNextBlockTimestamp", [
+      refundEndTime.toNumber() + 1,
+    ]);
+
+    await erc721RExample.provider.send("hardhat_setBalance", [
+      owner.address,
+      "0x6a94d74f430000", // 0.03 ether
+    ]);
+    const ownerOriginBalance = await erc721RExample.provider.getBalance(
+      owner.address
+    );
+    // first check the owner balance is less than 0.1 ether
+    expect(ownerOriginBalance).to.be.lt(parseEther("0.1"));
+
+    await erc721RExample.connect(owner).withdraw();
+
+    const contractVault = await erc721RExample.provider.getBalance(
+      erc721RExample.address
+    );
+    const ownerBalance = await erc721RExample.provider.getBalance(
+      owner.address
+    );
+
+    expect(contractVault).to.be.equal(parseEther("0"));
+    // the owner origin balance is less than 0.1 ether
+    expect(ownerBalance).to.be.gt(parseEther("0.1"));
+  });
+
+  /**
+   * Toggle Test
+   */
+  it("[Toggle] Owner can call toggleRefundCountdown and refundEndTime add `refundPeriod` days.", async function () {
+    const beforeRefundEndTime = (
+      await erc721RExample.refundEndTime()
+    ).toNumber();
+
+    await erc721RExample.provider.send("evm_setNextBlockTimestamp", [
+      beforeRefundEndTime,
+    ]);
+
+    await erc721RExample.toggleRefundCountdown();
+
+    const afterRefundEndTime = (
+      await erc721RExample.refundEndTime()
+    ).toNumber();
+
+    expect(afterRefundEndTime).to.be.equal(
+      beforeRefundEndTime + FORTY_FIVE_DAYS
+    );
+  });
+
+  it("[Toggle] Owner can call togglePresaleStatus", async function () {
+    await erc721RExample.togglePresaleStatus();
+    expect(await erc721RExample.presaleActive()).to.be.true;
+  });
+
+  it("[Toggle] Owner can call togglePublicSaleStatus", async function () {
+    await erc721RExample.togglePublicSaleStatus();
+    expect(await erc721RExample.publicSaleActive()).to.be.false;
+  });
+
+  /**
+   * Setter Test
+   */
+  it("[Setter] Owner can call setRefundAddress", async function () {
+    await erc721RExample.setRefundAddress(
+      "0x06f509F73eefBA36352Bc8228F9112C3786100dA"
+    );
+    expect(await erc721RExample.refundAddress()).to.be.equal(
+      "0x06f509F73eefBA36352Bc8228F9112C3786100dA"
+    );
+  });
+
+  it("[Setter] Owner can call setMerkleRoot", async function () {
+    await erc721RExample.setMerkleRoot(
+      "0x070e8db97b197cc0e4a1790c5e6c3667bab32d733db7f815fbe84f5824c7168d"
```

Comment:
for merkle roots we can do something along these lines (taken from some other code we have that isn't public yet):

when using a count:

```
  // based on: https://github.com/steve-ng/merkle-airdrop/blob/main/test/MerkleDistributor.ts#L9
    users = [
      { address: owner.address, amount: 10 },
      { address: account2.address, amount: 15 },
      { address: account3.address, amount: 20 },
      { address: account4.address, amount: 30 },
    ];
    elements = users.map((x) =>
      utils.solidityKeccak256(["address", "uint256"], [x.address, x.amount])
    );
    
    // Merkle tree util functions
function getTree(elements) {
  const tree = new MerkleTree(elements, keccak256, { sort: true });
  const root = tree.getHexRoot();
  return { tree, root };
}

function getProof(tree, leaf) {
  return tree.getHexProof(leaf);
}

    const leaf = elements[3];
    const amount = users[3].amount;
    const { root, tree } = getTree(elements);
    const proof = getProof(tree, leaf);

    await cryptoFightersPotion.setMerkleRoot(root);
```

when using just addresses you can skip this:

```
utils.solidityKeccak256(["address", "uint256"], [x.address, x.amount])
```

and just hash the address. something like this:

```
utils.solidityKeccak256(x.address)
```